### ISSUE TITLE
Can you explain these two lines?

### ISSUE THREAD

**Issue by darren1231** [elix-tech/dqn#2]
processed_observation = np.maximum(observation, last_observation)
processed_observation = np.uint8(resize(rgb2gray(processed_observation), (FRAME_WIDTH, FRAME_HEIGHT)) * 255)

I know everything but these two lines. Can you explain it detailly?
The first one I have saw some explanation on your blog, but it's written in Japanese.
I am still quite confused about it. 

And the second line, why you want to upside down the number?

This is my test in np.uint8 function. It looks like you want to upside down the number, doesn't it?
np.uint8(5*255)
Out[25]: 251

np.uint8(4*255)
Out[26]: 252

np.uint8(3*255)
Out[27]: 253

np.uint8(255*255)
Out[28]: 1

I have done this kind of environment for quite a long time, but I still can't get a success.
I have learned DQN from the basic example:https://github.com/asrivat1/DeepLearningVideoGames
I directly change the environment to Gym, it always failed. I thought I was wrong about the preprocess part. If you can give me some detailed explanation or document, I would be very much appreciated. ^.^

**Comment by tokb23** [comment]
`processed_observation = np.maximum(observation, last_observation)`
Here, it takes the maximum value for each pixel color value over the current frame and the previous frame. This is necessary to remove flickering that is present in games where some objects appear only in even frames while other objects appear only in odd frames, an artifact caused by the limited number of sprites Atari 2600 can display at once.

`resize(rgb2gray(processed_observation), (FRAME_WIDTH, FRAME_HEIGHT))`
This should output the value between 0.0-1.0 (the data type is float64) for each pixel value, so for instance,

np.uint8(0.1*255)
25

np.uint8(0.8*255)
204

and this just makes values between 0-255 so it seems fine (it doesn't turn values upside down).
I want to make the type of values uint8 because keeping float type of values for replay memory is expensive.

Since I only did the short experiment mentioned in my blog post, I didn't try it for longer time, but someone told me that even if running my code for quite a long time, at some point it stops improving.
For one reason, my implementation uses normal RMSProp but in the DQN paper the authors seem to use hacked RMSProp (you can also try Adam instead). And also, DQN seems sensitive to hyperparameters so you have to choose those more carefully.

Hope this helps.

**Comment by darren1231** [comment]
Thank you for your detailed explanation, it really solved my confusion.
By the way, what is the hyperparameters you mentioned about?
How many CNN layer? Or epsilon parameter? Or something else?

**Comment by tokb23** [comment]
I meant hyperparameters relating to optimizers, that is, learning rate, momentum, etc...
